src/core/ddsc/tests/psmx.c

```diff
@@ -1264,6 +1391,164 @@ CU_Test (ddsc_psmx, basic)
+  memset(locator_in, 0x0, sizeof(locator_in)); // Avoid warning 'uninitialized value'.
+  ((uint64_t*)locator_in)[0] = (uint64_t)0x4a4d203df6996395;
+  ((uint64_t*)locator_in)[1] = (uint64_t)0xe1412fbecc2de4b6;
+
+  int N = sizeof(specify_locator) / sizeof(specify_locator[0]);
+  for (int i = 0; i < N; ++i)
+  {
+    const uint8_t* locator = specify_locator[i] ? locator_in : NULL;
+    dds_entity_t participant = create_participant(domainId, locator);
+    CU_ASSERT_FATAL(participant > 0);
+    dds_entity_t domain = dds_get_parent(participant);
+    CU_ASSERT_FATAL(domain > 0);
+    if ( specify_locator[i] ) {
+      // Check that I get the same locator that I provided with the config.
+      ddsi_locator_t locator_out;
+      CU_ASSERT_FATAL(domain_get_psmx_locator(domain, &locator_out));
+      CU_ASSERT_FATAL(memcmp(locator_in, locator_out.address, sizeof(locator_out.address)) == 0);
+    }
+
+    {
+      // Query whether shared memory is supported.
+      dds_domain* dom = NULL;
+      CU_ASSERT_FATAL(domain_pin(domain, &dom));
+      CU_ASSERT_FATAL(dom->psmx_instances.length == 1); // There shall be exactly one psmx instance.
+      struct dummy_psmx* dpsmx = (struct dummy_psmx*)dom->psmx_instances.instances[0];
+      dpsmx->mockstats_get_ownership(&dmock);
+
+      // The psmx must have an instance_name that is not an empty string.
+      CU_ASSERT_FATAL(dpsmx->c.instance_name != NULL && strcmp(dpsmx->c.instance_name, "") != 0);
+      supports_shared_memory = ((dpsmx->c.ops.supported_features(&dpsmx->c) & DDS_PSMX_FEATURE_SHARED_MEMORY) == DDS_PSMX_FEATURE_SHARED_MEMORY);
+      domain_unpin(dom);
+    }
+    dds_entity_t writer1, reader1, writer2, reader2;
+    {
+      char topicname[100];
+      dds_qos_t* qos = dds_create_qos();
+
+      create_unique_topic_name("shared_memory", topicname, sizeof(topicname));
+      dds_entity_t topic1 = dds_create_topic(participant, &SC_Model_desc, topicname, qos, NULL);
+      CU_ASSERT_FATAL(topic1 > 0);
+
+      writer1 = dds_create_writer(participant, topic1, qos, NULL);
+      CU_ASSERT_FATAL(writer1 > 0);
+
+      reader1 = dds_create_reader(participant, topic1, qos, NULL);
+      CU_ASSERT_FATAL(reader1 > 0);
+
+      create_unique_topic_name("shared_memory", topicname, sizeof(topicname));
+      dds_entity_t topic2 = dds_create_topic(participant, &PsmxType1_desc, topicname, qos, NULL);
+      CU_ASSERT_FATAL(topic2 > 0);
+
+      writer2 = dds_create_writer(participant, topic2, qos, NULL);
+      CU_ASSERT_FATAL(writer2 > 0);
+
+      reader2 = dds_create_reader(participant, topic2, qos, NULL);
+      CU_ASSERT_FATAL(reader2 > 0);
+      dds_delete_qos(qos);
+    }
+
+    {
+      // Check that shared memory is enabled when it should, and not enabled when it shouldn't.
+      bool psmx_enabled;
+      psmx_enabled = endpoint_has_psmx_enabled(writer1);
+      CU_ASSERT_FATAL(psmx_enabled);
+      CU_ASSERT_FATAL(dds_is_shared_memory_available(writer1) == supports_shared_memory);
```

Comment:
So here is the aforementioned test for `supported_features` already. Of course you do have to run through this code *twice*, because otherwise `dds_is_shared_memory_available` could get away with returning a constant.